src/extensions/nuget/Microsoft.DotNet.UpgradeAssistant.Extensions.NuGet/NuGetTransitiveDependencyIdentifier.cs

```diff
@@ -0,0 +1,199 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.UpgradeAssistant.Dependencies;
+using Microsoft.Extensions.Logging;
+using NuGet.Commands;
+using NuGet.Configuration;
+using NuGet.DependencyResolver;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectModel;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.NuGet
+{
+    public class NuGetTransitiveDependencyIdentifier : ITransitiveDependencyIdentifier
+    {
+        private readonly IEnumerable<PackageSource> _packageSources;
+        private readonly ISettings _settings;
+        private readonly SourceCacheContext _context;
+        private readonly ILogger<NuGetTransitiveDependencyIdentifier> _logger;
+
+        public NuGetTransitiveDependencyIdentifier(
+            IEnumerable<PackageSource> packageSources,
+            ISettings settings,
+            SourceCacheContext context,
+            ILogger<NuGetTransitiveDependencyIdentifier> logger)
+        {
+            _packageSources = packageSources ?? throw new ArgumentNullException(nameof(packageSources));
+            _settings = settings ?? throw new ArgumentNullException(nameof(settings));
+            _context = context ?? throw new ArgumentNullException(nameof(context));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<TransitiveClosureCollection> GetTransitiveDependenciesAsync(IEnumerable<NuGetReference> packages, IEnumerable<TargetFrameworkMoniker> tfms, CancellationToken token)
+        {
+            if (packages is null)
+            {
+                throw new ArgumentNullException(nameof(packages));
+            }
+
+            if (tfms is null)
+            {
+                throw new ArgumentNullException(nameof(tfms));
+            }
+
+            var graph = await RestoreProjectAsync(packages, tfms, token).ConfigureAwait(false);
+
+            if (graph is null)
+            {
+                return TransitiveClosureCollection.Empty;
+            }
+
+            return new(new TargetGraphLookup(graph));
+        }
+
+        private async Task<RestoreTargetGraph?> RestoreProjectAsync(IEnumerable<NuGetReference> packages, IEnumerable<TargetFrameworkMoniker> tfms, CancellationToken token)
+        {
+            var tfmInfo = tfms.Select(tfm => new TargetFrameworkInformation { FrameworkName = NuGetFramework.Parse(tfm.ToFullString()) }).ToList();
+
+            // Create a project in a unique and temporary directory
+            var path = Path.Combine(Path.GetTempPath(), "dotnet-ua", "restores", Guid.NewGuid().ToString(), "project.txt");
+
```

Comment:
Since we are locally doing the restore and cleanup transitive dependencies, I'm expecting this to be much more performant than needing to do `dotnet restore`, is there a cleanup strategy for the temp folder structure created? also will this be an issue with larger projects?

Reply:
good question. since it's in the temp path I imagine it will just be cleaned up by the os at some point. I'll create an issue to track that.